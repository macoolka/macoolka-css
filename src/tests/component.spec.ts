
import { standRuleModule, typeThemeLens, StandRule, ExtendRule, RuleModule, GetRuleProp,
      parse as _parse,parseProp as _parseProp } from '..';
const theme:any=standRuleModule.theme;
const darkTheme:any = typeThemeLens.set('dark')(theme);
/**
 * Enum Properties's value using string or number,
 * The tag 'ruleEnum' in rule describe the property style
 */
export type EProp = {
    size?: 'small' | 'medium'
};
/**
 * Stand Properties's value using any type,
 * The tag 'rule' in rule describe the property style
 */
export type SProp = {
    disabled?: boolean
};
/**
 * The output props. including predefine property and orginal property
 */
export type Props = GetRuleProp<Rule>

const rule: Rule = {
    /**
     * default style
     */
    style: {
        mkFontSize: 'h3',
        position: 'static',
        selector: {
            ':focus': {
                width: 30,
            },
        },
    },
    /**
     * descirbe stand property.
     * the type is
     * propertyName :({value,name,theme,source})=>cssnode
     * value is current value.
     * name is propertyname
     * theme is gloabe theme
     * source is component props
     */
    rule: {
        disabled: ({ value }) => ({
            opacity: value ? 0.38 : 1,
        }),
    },
    /**
     * descirbe enum property.
     * the type is
     * propertyName :{propertyvalue:({value,name,theme,source})=>cssnode}
     */
    ruleEnum: {
        size: {
            medium: () => ({
                mkFontSize: 'h2',
                mkMedia: [{ mkPadding: 'small' }, { mkPadding: 'medium' }],
                selector: {
                    ':focus': {
                        width: 10,
                    },
                },
            }),
            small: () => ({
                mkFontSize: 'h1',
                mkHoverTextColor: 'accent',
                selector: {
                    ':focus': {
                        width: 10,
                    },
                },
            }),


        },
    },
};
export type Rule = ExtendRule<StandRule, SProp, EProp>
export const _ruleModule: RuleModule<Rule> = {
    rule,
    theme,
    next:standRuleModule
}
export const parse = _parse(_ruleModule);
export const parseDark = _parse({ ..._ruleModule, theme: darkTheme });
describe('should parse theme', () => {
    it('should parse light theme', () => {
        expect(parse({
            mkColor: 'main',
        })).toMatchSnapshot()
    });
    it('should parse dark theme', () => {
        expect(parseDark({
            mkColor: 'main',
          
        })).toMatchSnapshot()
    });
    it('should parse dark with prop theme', () => {
        expect(parseDark({
            mkColor: 'main',
            theme: {
                color: {
                    type: 'dark'
                }
            } as any
        })).toMatchSnapshot()
    });
});
describe('should parse component rule', () => {
    it('1 should parse empty props', () => {
        expect(parse({
        })).toMatchSnapshot()
    });
    it('2 should parse common props', () => {
        expect(parse({
            height: 180,
        })).toMatchSnapshot()
    });
    it('3 should parse common props for place rule props', () => {
        expect(parse({
            fontSize: 30,
        })).toMatchSnapshot()
    });
    it('4 should parse component props for palce default style ', () => {
        expect(parse({
            size: 'small',
        })).toMatchSnapshot()
    });
    it('5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style', () => {
        expect(parse({
            fontSize: 7,
            mkFontSize: 'p',
            mkTypography: 'h1',
            size: 'small',
        })).toMatchSnapshot()
        expect(parse({
            mkFontSize: 'p',
            mkTypography: 'h1',
            size: 'small',
        })).toMatchSnapshot()
        expect(parse({
            mkFontSize: 'overline',
            size: 'small',
        })).toMatchSnapshot()
        expect(parse({
            size: 'small',
        })).toMatchSnapshot()
        expect(parse({
        })).toMatchSnapshot()
    });
    it('6 should parse media props', () => {
        expect(parse({
            mkMedia: [{ mkPadding: 'small' }, { mkPadding: 'medium' }],
        })).toMatchSnapshot();
    });
    it('7 should parse media props in rule', () => {
        expect(parse({
            size: 'medium',
        })).toMatchSnapshot();
    });
    it('8 should parse advance props', () => {
        expect(parse({
            mkAbsolute: 'bottom',
        })).toMatchSnapshot();
    });
});


describe('should compose rule', () => {
    type Rule1 = ExtendRule<StandRule, {}, { a?: 'a' | 'b' }>
    type Rule2 = ExtendRule<Rule1, {}, { b?: 'a' | 'b' }>

   // type Rule = ExtendRule<StandRule, {}, { a?: 'a' | 'b' } & { b?: 'a' | 'b' }>
    const input1: RuleModule<Rule1> = {
        rule: {
            ruleEnum: {
                a: {
                    a: {
                        color: 'red'
                    },
                    b: {
                        color: 'green',
                    }
                }
            }
        },
        theme: theme,
        next:standRuleModule
    }
    const input2: RuleModule<Rule2> = {
        rule: {
            ruleEnum: {
                b: {
                    a: {
                        backgroundColor: 'yellow'
                    },
                    b: {
                        backgroundColor: 'pink',
                    }
                }
            }
        },
        theme: theme,
        next:input1
    }
    const parseProp = _parseProp(input2);

    it('simple rule', () => {
        expect(parseProp({
            a: 'a', b: 'a'
        })).toEqual({
            color: 'red',
            backgroundColor: 'yellow',
        })
    });

}); 