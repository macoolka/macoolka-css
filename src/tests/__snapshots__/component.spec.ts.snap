// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse component rule 1 should parse empty props 1`] = `
"font-size: 48px;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse component rule 2 should parse common props 1`] = `
"font-size: 48px;
height: 180px;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse component rule 3 should parse common props for place rule props 1`] = `
"font-size: 30px;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse component rule 4 should parse component props for palce default style  1`] = `
"font-size: 96px;
position: static;
:focus {
  width: 10px;
}
:hover:not(:disabled) {
  color: #f50057;
}
"
`;

exports[`should parse component rule 5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style 1`] = `
"font-size: 7px;
letter-spacing: -0.01562em;
font-weight: 300;
font-family: Roboto, Open Sans, Helvetica Neue, Helvetica, Arial, sans-serif;
line-height: 6rem;
text-decoration: inherit;
text-transform: inherit;
position: static;
:focus {
  width: 10px;
}
:hover:not(:disabled) {
  color: #f50057;
}
"
`;

exports[`should parse component rule 5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style 2`] = `
"font-size: 14px;
letter-spacing: -0.01562em;
font-weight: 300;
font-family: Roboto, Open Sans, Helvetica Neue, Helvetica, Arial, sans-serif;
line-height: 6rem;
text-decoration: inherit;
text-transform: inherit;
position: static;
:focus {
  width: 10px;
}
:hover:not(:disabled) {
  color: #f50057;
}
"
`;

exports[`should parse component rule 5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style 3`] = `
"font-size: 10px;
position: static;
:focus {
  width: 10px;
}
:hover:not(:disabled) {
  color: #f50057;
}
"
`;

exports[`should parse component rule 5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style 4`] = `
"font-size: 96px;
position: static;
:focus {
  width: 10px;
}
:hover:not(:disabled) {
  color: #f50057;
}
"
`;

exports[`should parse component rule 5 should parse common props and component props using the order base prop>module prop > advance prop>component prop >component default style 5`] = `
"font-size: 48px;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse component rule 6 should parse media props 1`] = `
"font-size: 48px;
position: static;
:focus {
  width: 30px;
}
@media screen and (max-width: 80em) {
  padding: 16px;
}
@media screen and (max-width: 50em) {
  padding: 8px;
}
"
`;

exports[`should parse component rule 7 should parse media props in rule 1`] = `
"font-size: 60px;
position: static;
:focus {
  width: 10px;
}
@media screen and (max-width: 80em) {
  padding: 16px;
}
@media screen and (max-width: 50em) {
  padding: 8px;
}
"
`;

exports[`should parse component rule 8 should parse advance props 1`] = `
"bottom: 0px;
font-size: 48px;
left: 0px;
right: 0px;
width: 100%;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse theme should parse dark theme 1`] = `
"font-size: 48px;
background-color: #00796b;
color: #fff;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse theme should parse dark with prop theme 1`] = `
"font-size: 48px;
background-color: #00796b;
color: #fff;
position: static;
:focus {
  width: 30px;
}
"
`;

exports[`should parse theme should parse light theme 1`] = `
"font-size: 48px;
background-color: #26a69a;
color: #fff;
position: static;
:focus {
  width: 30px;
}
"
`;
